distsql/distsql_test.go

```diff
@@ -43,3 +50,57 @@ func (s *testTableCodecSuite) TestColumnToProto(c *C) {
 	pc := columnToProto(col)
 	c.Assert(pc.GetFlag(), Equals, int32(10))
 }
+
+// For Issue 1791
+func (s *testTableCodecSuite) TestGoroutineLeak(c *C) {
+	var sr SelectResult
+	countBefore := runtime.NumGoroutine()
+
+	sr = &selectResult{
+		resp:    &mockResponse{},
+		results: make(chan PartialResult, 5),
+		done:    make(chan error, 1),
+		closed:  make(chan struct{}),
+	}
+	go sr.Fetch()
+	for {
+		// mock test will generate 500 partial result then return error
+		_, err := sr.Next()
+		if err != nil {
+			// close selectResult on error, partialResult's fetch goroutine may leak
+			sr.Close()
+			break
+		}
+	}
+
+	time.Sleep(3 * time.Second)
```

Comment:
3 seconds seems too long for a single test case.
We can wait a few mili-seconds in a for loop, once the count has reduced to 5, we can break directly.